File: zul/src/org/zkoss/zul/Progressmeter.java
/* Progressmeter.java

	Purpose:
		
	Description:
		
	History:
		Mon Aug 14 16:38:24     2006, Created by tomyeh

Copyright (C) 2006 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under LGPL Version 2.1 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
*/
package org.zkoss.zul;

import java.io.IOException;

import org.zkoss.zk.ui.UiException;
import org.zkoss.zul.impl.XulElement;

/**
 * A progress meter is a bar that indicates how much of a task has been completed. 
 *
 * <p>Default {@link #getZclass}: z-progressmeter. (since 3.5.0)
 *
 * @author tomyeh
 */
public class Progressmeter extends XulElement {
	private int _val;
	private boolean _resetWidth = true; //B80-ZK-2895

	public Progressmeter() {
		super.setWidth("100px");
	}
	public Progressmeter(int value) {
		this();
		setValue(value);
	}

	/** Overrides the method in HtmlBasedComponent, to avoid misuse hflex and width at the same time.
	 * @since 8.0.1
	 */
	@Override
	public void setWidth(String width) { //B80-ZK-2895
		_resetWidth = false;
		super.setWidth(width);
	}
	/** Overrides the method in HtmlBasedComponent, to avoid misuse hflex and width at the same time.
	 * @since 8.0.1
	 */
	@Override
	public void setHflex(String flex) { //B80-ZK-2895
		if (_resetWidth)
			super.setWidth("");
		super.setHflex(flex);
	}
	/** Sets the current value of the progress meter.
	 * <p>Range: 0~100.
	 */
	public void setValue(int value) {
		if (value < 0 || value > 100)
			throw new UiException("Illegal value: "+value+". Range: 0 ~ 100");
		if (_val != value) {
			_val = value;
			smartUpdate("value", _val);
		}
	}
	/** Returns the current value of the progress meter.
	 */
	public int getValue() {
		return _val;
	}

	//-- super --//
	public String getZclass() {
		return _zclass == null ? "z-progressmeter" : _zclass;
	}

	//-- Component --//
	protected boolean isChildable() {
		return false;
	}
	
	//super//
	protected void renderProperties(org.zkoss.zk.ui.sys.ContentRenderer renderer)
	throws IOException {
		super.renderProperties(renderer);
		render(renderer, "value", ""+_val);
		
	}
}

File: zul/src/main/resources/web/js/zul/wgt/Progressmeter.ts
/* Progressmeter.ts

	Purpose:

	Description:

	History:
		Thu May 14 10:17:24     2009, Created by kindalu

Copyright (C) 2009 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.1 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/
/**
 * A progress meter is a bar that indicates how much of a task has been completed.
 *
 * @defaultValue {@link getZclass}: z-progressmeter.
 */
@zk.WrapClass('zul.wgt.Progressmeter')
export class Progressmeter extends zul.Widget {
	/** @internal */
	_value = 0;
	/** @internal */
	_indeterminate = false;
	/** @internal */
	_indeterminateAnimation = false;

	/**
	 * @returns the current value of the progress meter.
	 */
	getValue(): number {
		return this._value;
	}

	/**
	 * Sets the current value of the progress meter.
	 * <p>Range: 0~100.
	 */
	setValue(value: number, opts?: Record<string, boolean>): this {
		const o = this._value;
		this._value = value;

		if (o !== value || opts?.force) {
			if (this.$n())
				this._fixImgWidth();
		}

		return this;
	}

	/**
	 * @returns the indeterminate state of the progress meter.(default false)
	 */
	isIndeterminate(): boolean {
		return this._indeterminate;
	}

	/**
	 * Sets the indeterminate state of the progress meter.
	 */
	setIndeterminate(indeterminate: boolean, opts?: Record<string, boolean>): this {
		const o = this._indeterminate;
		this._indeterminate = indeterminate;

		if (o !== indeterminate || opts?.force) {
			if (this.$n()) {
				jq(this.$n()).toggleClass(this.$s('indeterminate'), indeterminate);
			}
		}

		return this;
	}

	/** @internal */
	_fixImgWidth(): void {
		var n = this.$n(),
			img = this.$n('img');
		if (img) {
			//B70-ZK-2453 remember to add brackets
			if (zk(n).isRealVisible() && !this._indeterminateAnimation) { //Bug 3134159
				var $img = jq(img);
				$img.animate({
					width: this._value + '%'
				}, { duration: $img.zk.getAnimationSpeed('slow'), queue: false, easing: 'linear' }); //ZK-4079: progressmeter animation not catching up with actual value
			}
		}
	}

	override onSize(): void {
		this._fixImgWidth();
	}

	/** @internal */
	override bind_(desktop?: zk.Desktop, skipper?: zk.Skipper, after?: CallableFunction[]): void {//after compose
		super.bind_(desktop, skipper, after);
		this._fixImgWidth();
		zWatch.listen({onSize: this});
	}

	/** @internal */
	override unbind_(skipper?: zk.Skipper, after?: CallableFunction[], keepRod?: boolean): void {
		zWatch.unlisten({onSize: this});
		super.unbind_(skipper, after, keepRod);
	}

	override setWidth(width?: string): this {
		super.setWidth(width);
		this._fixImgWidth();
		return this;
	}

	/** @internal */
	override domClass_(no?: zk.DomClassOptions): string {
		var scls = super.domClass_(no);
		if (!no || !no.zclass) {
			if (this._indeterminate)
				scls += ' ' + this.$s('indeterminate');
		}
		return scls;
	}
}